File: src/util/activity-tracking.ts
import {
    readActivityTrackingConfig,
    writeActivityTrackingConfig
} from ':util/data.ts';
import { log } from ':util/logger.ts';
import { Activity, Member } from 'discordeno';

/**
 * Checks for Lunaro in a list of activities. Looks for both Warframe itself and
 * [PyLunaroRPC](https://github.com/kozabrada123/PyLunaroRPC)
 *
 * @param activities to be checked
 * @returns whether any of the activities is Lunaro
 */
export const doActivitiesIncludeLunaro = (activities: Activity[]) => {
    const localizedLunaroName = ['lunaro', 'лунаро', '루나로'];

    const lunaroActivities = activities.filter((activity) => {
        const isWarframeLunaro =
            activity.name.toLowerCase() === 'warframe' &&
            localizedLunaroName.includes(activity.details?.toLowerCase() || '');

        const isPyLunaroRPC =
            activity.name.toLowerCase() === 'warframe: lunaro';

        return isWarframeLunaro || isPyLunaroRPC;
    });

    return lunaroActivities.length > 0;
};

/** Updates the activity tracking config to allow tracking globally. */
export const enableActivityTracking = () => {
    const data = readActivityTrackingConfig();
    data.enabled = true;
    writeActivityTrackingConfig(data);

    log('Enabled activity tracker');
};

/** Updates the activity tracking config to deny tracking globally. */
export const disableActivityTracking = () => {
    const data = readActivityTrackingConfig();
    data.enabled = false;
    writeActivityTrackingConfig(data);

    log('Disabled activity tracker');
};

/**
 * Adds a member to the activity tracking blocklist to disable tracking for
 * their account's activity.
 *
 * @param member to be blocklisted
 */
export const addMemberToTrackingBlocklist = (member: Member) => {
    const data = readActivityTrackingConfig();

    if (!data.blocklist.includes(member.id.toString())) {
        data.blocklist.push(member.id.toString());
        writeActivityTrackingConfig(data);

        log('Member added to activity tracking blocklist');
    }
};

/**
 * Removes a member to the activity tracking blocklist to resume tracking for
 * their account's activity.
 *
 * @param member to be blocklisted
 */
export const removeMemberFromTrackingBlocklist = (member: Member) => {
    const data = readActivityTrackingConfig();

    if (data.blocklist.includes(member.id.toString())) {
        data.blocklist = data.blocklist.filter(
            (entry) => entry !== member.id.toString()
        );
        writeActivityTrackingConfig(data);

        log('Member removed from activity tracking blocklist');
    }
};

File: src/util/activity_tracking.rs
use std::{collections::HashSet, sync::OnceLock};

use async_trait::async_trait;
use futures::lock::{Mutex, MutexGuard};
use poise::serenity_prelude::{Presence, User};
use serde::{Deserialize, Serialize};

use crate::{errors::data::DataError, traits::config_file::ConfigFile, types::error::Error};

use super::data;

static CONFIG: OnceLock<Mutex<Config>> = OnceLock::new();
const CONFIG_FILE: &str = "activity_tracking.json";

/// Configures the activity tracker's behaviour.
#[derive(Serialize, Deserialize, Default, Debug)]
pub struct Config {
    /// List of user IDs to ignore activity updates from.
    pub blocklist: HashSet<u64>,
}

impl Config {
    /// Check if a user is on the tracking blocklist.
    pub fn is_blocked(&self, user_id: u64) -> bool {
        self.blocklist.contains(&user_id)
    }

    /// Add a user to the tracking blocklist.
    fn add_to_blocklist(&mut self, user_id: u64) -> Result<(), Error> {
        self.blocklist.insert(user_id);
        self.save()
    }

    /// Remove a user from the tracking blocklist.
    fn remove_from_blocklist(&mut self, user_id: u64) -> Result<(), Error> {
        self.blocklist.remove(&user_id);
        self.save()
    }
}

#[async_trait]
impl ConfigFile for Config {
    fn load() -> Result<Box<Self>, Error> {
        match data::read_config(CONFIG_FILE) {
            Ok(config) => Ok(config),
            Err(error) => match error.downcast_ref::<DataError>() {
                Some(DataError::MissingConfigFile(_)) => {
                    let config = Config::default();
                    config.save()?;
                    Ok(config.into())
                }
                _ => Err(error),
            },
        }
    }

    fn save(&self) -> Result<(), Error> {
        data::write_config(CONFIG_FILE, self)
    }

    async fn instance<'a>() -> MutexGuard<'a, Self> {
        CONFIG
            .get_or_init(|| Mutex::new(*Config::load().unwrap_or_default()))
            .lock()
            .await
    }
}

/// Remove a user from the tracking blocklist, if present.
pub async fn allow_for(user: &User) -> Result<(), Error> {
    let mut config = Config::instance().await;

    config.remove_from_blocklist(user.id.0)?;

    log::debug!(
        "Removed {} ({}) from tracking blocklist",
        user.tag(),
        user.id
    );

    Ok(())
}

/// Add a user to the tracking blocklist.
pub async fn deny_for(user: &User) -> Result<(), Error> {
    let mut config = Config::instance().await;

    config.add_to_blocklist(user.id.0)?;

    log::debug!("Added {} ({}) to tracking blocklist", user.tag(), user.id);
    Ok(())
}

/// Check if a presence update includes Lunaro activity.
pub fn is_playing_lunaro(presence: &Presence) -> Result<bool, Error> {
    let localized_lunaro = ["lunaro", "лунаро", "루나로"];

    let is_lunaro = presence.activities.iter().any(|activity| {
        if let Some(mission) = &activity.details {
            activity.name.to_lowercase() == "warframe"
                && localized_lunaro.contains(&mission.to_lowercase().as_ref())
        } else {
            false
        }
    });

    // https://github.com/kozabrada123/PyLunaroRPC
    // https://github.com/imatpot/lunaro-manager/pull/1
    let is_py_lunaro_rpc = presence
        .activities
        .iter()
        .any(|activity| activity.name.to_lowercase() == "warframe: lunaro");

    Ok(is_lunaro || is_py_lunaro_rpc)
}
